### PR TITLE
Handle Pinot Project Expressions - SWITCH

### PR DESCRIPTION
**Pull request by cliandy**
Support Presto switch statement pushdown to Pinot. The motivation for this is to allow for less data transfer when working with multicolumn projections like CASE statements, other functionality, and leading up to supporting COALESCE once NULL can be properly handled.

Enable this feature by setting `pinot.pushdown_project_expressions=true`

Test plan
- Modified Unit Tests
- Explain Queries (i.e.)
```
GeneratedPinotQuery{query=SELECT CASE true WHEN ("numberOfGames" > "numberOfGamesAsBatter") THEN "playerName" ELSE "teamID" END FROM baseballStats__TABLE_NAME_SUFFIX_TEMPLATE____TIME_BOUNDARY_FILTER_TEMPLATE__ LIMIT 2147483647
```

```
== RELEASE NOTES ==

Pinot Changes
* Add support for `CASE` statement pushdown to Pinot
* Add support for logical and arithmetic functions for `CASE` statements.
```

### CODE REVIEW

**Comment by cliandy** [comment]
@xiangfu0 let me know your thoughts when you get a chance

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java [hunk 1]
```diff
@@ -78,8 +101,29 @@ protected boolean isImplicitCast(Type inputType, Type resultType)
...
+            PinotExpression right = arguments.get(1).accept(this, context);
+            String prestoOperator = operatorType.getOperator();
+            String pinotOperator = PRESTO_TO_PINOT_OPERATORS.get(prestoOperator);
+            if (pinotOperator == null) {
+                throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), "Unsupported binary expression " + prestoOperator);
+            }
+            return derived(format("%s(%s, %s)", pinotOperator, left.getDefinition(), right.getDefinition()));
+        }
+        throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), format("Don't know how to interpret %s as an arithmetic expression", expression));
+    }
```
Comment (xiangfu0) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java:125]: nit: newline

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java [hunk 2]
```diff
@@ -28,30 +33,48 @@
...
 import static com.facebook.presto.pinot.PinotErrorCode.PINOT_UNSUPPORTED_EXPRESSION;
+import static com.facebook.presto.pinot.PinotPushdownUtils.getLiteralAsString;
+import static com.facebook.presto.pinot.query.PinotExpression.derived;
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
 class PinotProjectExpressionConverter
         implements RowExpressionVisitor<PinotExpression, Map<VariableReferenceExpression, Selection>>
 {
+    private static final Set<String> LOGICAL_BINARY_OPS_FILTER = ImmutableSet.of("=", "<", "<=", ">", ">=", "<>");
```
Comment (xiangfu0) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java:54]: Pinot uses `!=` for not equal, you may need to convert it when generating a pinot query.
Or you can use the map to make it generic just like this `PRESTO_TO_PINOT_OPERATORS`.

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java [hunk 3]
```diff
@@ -78,8 +101,29 @@ protected boolean isImplicitCast(Type inputType, Type resultType)
...
-    private PinotExpression handleCast(
+    protected PinotExpression handleArithmeticExpression(
+            CallExpression expression,
+            OperatorType operatorType,
+            Map<VariableReferenceExpression, PinotQueryGeneratorContext.Selection> context)
+    {
+        List<RowExpression> arguments = expression.getArguments();
+        if (arguments.size() == 1) {
+            String prefix = operatorType == OperatorType.NEGATION ? "-" : "";
+            return derived(prefix + arguments.get(0).accept(this, context).getDefinition());
```
Comment (xiangfu0) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java:112]: can you give some examples for these negation cases?
Will the negation sign just work or the NOT statement is required?

**Review by xiangfu0 — COMMENTED** [review]
Overall LGTM! 
Some generated query details should be tested for pinot syntax, mostly negation and not equal.

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java (hunk 2 above)
Comment (cliandy) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotProjectExpressionConverter.java:54]: my understanding here is that based on the calcite spec, `<>` and `!=` are both okay with `!=` permissible based on the level of conformance, which is currently at BABEL. Tested `<>` and this runs well, so curious where this is defined internally?

**Comment by cliandy** [comment]
@NikhilCollooru can you take a look at this when you get a chance? Thanks!

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java [hunk 4]
```diff
@@ -373,20 +371,42 @@ public PinotExpression visitVariableReference(VariableReferenceExpression refere
         Selection input = requireNonNull(context.apply(reference), format("Input column %s does not exist in the input: %s", reference, context));
         return new PinotExpression(input.getDefinition(), input.getOrigin());
     }
-
+    private String getExpressionOrConstantString(RowExpression expression,
+                                                 Function<VariableReferenceExpression, Selection> context)
```
Comment (highker) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java:375]: put into the same line

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java [hunk 5]
```diff
@@ -373,20 +371,42 @@ public PinotExpression visitVariableReference(VariableReferenceExpression refere
...
+    {
+        if (expression instanceof ConstantExpression) {
+            return new PinotExpression(getLiteralAsString((ConstantExpression) expression),
```
Comment (highker) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java:378]: put the first param to its own line

**Review by highker — APPROVED** [review]
could you squash two commits into one?

**Comment by cliandy** [comment]
@highker let me know if there are any other concerns, thanks!

**Comment by highker** [comment]
Test failure is related

**Comment by highker** [comment]
Could squash all commits into one?

**Pull request merged**